tripy/examples/segment-anything-model-v2/video_demo.py

```diff
@@ -0,0 +1,178 @@
+
+# This source code is licensed under the license found in the
+# LICENSE file in the root directory of this source tree.
+
+# Not a contribution
+# Changes made by NVIDIA CORPORATION & AFFILIATES enabling SAM2 with Tripy or otherwise documented as
+# NVIDIA-proprietary are not a contribution and subject to the following terms and conditions:
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+from sam2.build_sam import build_sam2_video_predictor
+import numpy as np
+import matplotlib.pyplot as plt
+from PIL import Image
+from demo_utils import process_and_show_mask as show_mask
+from typing import Optional
+
+import os
+import time
+
+
+def compute_mask_properties(mask):
+    # Ensure we have a boolean array
+    test_mask = np.asarray(mask, dtype=bool)
+
+    # Calculate basic stats
+    volume = np.sum(test_mask)
+
+    # Calculate centroid (center of mass)
+    indices = np.where(test_mask)
+    assert volume > 0
+    centroid = tuple(float(np.mean(idx)) for idx in indices)
+    return volume, centroid
+
+
+def main(video_dir: str, save_path: Optional[str] = None):
+    """
+    Main execution function.
+
+    Args:
+        video_path (str): Path to where video frames are stored
+        save_path (str, optional): Directory to save visualizations
+
+    Returns:
+        Dict[str, np.ndarray]: Processing results
+    """
+
+    sam2_checkpoint = "./checkpoints/sam2.1_hiera_large.pt"
+    model_cfg = "sam2_hiera_l.yaml"
+    predictor = build_sam2_video_predictor(model_cfg, sam2_checkpoint, device=torch.device("cuda"))
+
+    # scan all the JPEG frame names in this directory
+    frame_names = [p for p in os.listdir(video_dir) if os.path.splitext(p)[-1] in [".jpg", ".jpeg", ".JPG", ".JPEG"]]
+    frame_names.sort(key=lambda p: int(os.path.splitext(p)[0]))
+    # take a look the first video frame
+    frame_idx = 0
+    if save_path:
+        plt.figure(figsize=(9, 6))
+        plt.title(f"frame {frame_idx}")
+        plt.imshow(Image.open(os.path.join(video_dir, frame_names[frame_idx])))
+        plt.savefig(os.path.join(save_path, f"video_{frame_idx}.png"))
+        plt.close("all")
+
+    inference_state = predictor.init_state(video_path=video_dir)
+
+    def make_tensors_contiguous(d):
+        for key, value in d.items():
+            if isinstance(value, torch.Tensor):
```

Comment:
Here too, why can the type sometimes be a torch tensor and other times not?

Reply:
Not all entries in `dict` are Tensors.